Clippy fixes

XAMPPRocky opened the pull request:
This PR fixes various clippy lints when building the `wasmtime` package.

sunfishcode commented on crates/environ/src/vmoffsets.rs:
```diff
@@ -19,7 +19,7 @@ fn cast_to_u32(sz: usize) -> u32 {
         Ok(x) => x,
-        Err(_) => panic!("overflow in cast from usize to u32"),
+        _ => panic!("overflow in cast from usize to u32"),
     }
```
This whole `match` could be simplified to a `.expect` call.

sunfishcode commented on crates/environ/src/compilation.rs:
```diff
@@ -66,6 +66,11 @@ impl Compilation {
 
+    /// Returns whether there are no functions defined.
+    pub fn is_empty(&self) -> bool {
+        self.functions.len() == 0
```
This should call `self.functions.is_empty()` too.

sunfishcode commented on crates/environ/src/func_environ.rs:
```diff
@@ -450,9 +450,9 @@ impl<'module_environment> cranelift_wasm::FuncEnvironment for FuncEnvironment<'m
-                memory: _,
                 style: MemoryStyle::Dynamic,
                 offset_guard_size,
+                ..
```
If someone adds a new field to `MemoryPlan`, the old code here will oblige them to look at this code and figure out if it needs to think about the new field. With `..`, this code will just silently ignore the new field. There are cases where `..` makes sense, when we're reasonably sure we won't care about other fields which may be added in the future, but I don't think that's the case in the code here.

sunfishcode commented on crates/environ/src/func_environ.rs:
```diff
@@ -469,9 +469,9 @@ impl<'module_environment> cranelift_wasm::FuncEnvironment for FuncEnvironment<'m
-                memory: _,
                 style: MemoryStyle::Static { bound },
                 offset_guard_size,
+                ..
```
Ditto.

sunfishcode commented on crates/environ/src/cache/worker.rs:
```diff
@@ -319,93 +321,106 @@ impl WorkerThread {
+                    }
+                }
+            };
+        }
```
All uses of `unwrap_or_warn` do `error => warn!(...)`. Would it be difficult to fold that into the macro?

Would it make sense to generalize the `unwrap_or` macro used elsewhere in this file and use that here?

XAMPPRocky replied:
Well I initially decided to have it contain `error => <expr>` so that if someone wanted to change the warning logging they wouldn't need to change the macro. If merging the two macros having the same logging in both `unwrap_or` places is okay with the project then I can make that change.

sunfishcode replied:
Yeah. Unless there's some code that differs that I missed, I think it makes sense to use `unwrap_or` consistently through the code here.

XAMPPRocky commented:
@sunfishcode Thanks for your feedback I've implemented it now, and while I was doing that clippy decided to output more lints so I've also fixed those where possible.  There are now three main categories of clippy lints that I didn't solve with this problem. "Missing Safety" documentation, "Type is too complex", and "Function takes too many arguments".

For adding the safety documentation for most of the API I figured it would be better as a separate PR. For the other two I wanted to know if there were opinions on these two lints? How many function arguments is too many? Should someone use type alias for complex return types?

sunfishcode commented:
For the number of arguments or complexity of return types, I think it's best to focus on "is there any way we could restructure the code to make it more readable?" rather than on following any absolute rule. It'll be subjective, but that's ok. If we can't find a way to make the code avoid those kinds of lints without making it less readable, I think it's ok to ignore and/or suppress clippy as needed.

sunfishcode commented on crates/api/src/module.rs:
```diff
@@ -24,6 +24,7 @@ fn into_global_type(gt: &wasmparser::GlobalType) -> GlobalType {
 }
 
+#[allow(clippy::trivially_copy_pass_by_ref)]
 fn into_valtype(ty: &wasmparser::Type) -> ValType {
```
`wasmparser::Type` implements `Copy` and is a simple enum; is there a reason for passing it by reference here?

XAMPPRocky replied:
@sunfishcode this function is used with `map` which requires `&T`.

sunfishcode replied:
Ah, makes sense.

alexcrichton commented:
FWIW I figure it'd worth posting my experience with clippy here, and unfortunately it's pretty poor. We've gotten [dozens of PRs](https://github.com/rust-lang/cargo/pulls?utf8=%E2%9C%93&q=is%3Apr+clippy+) on Cargo fixing various clippy warnings, and I find that they tend to fall in a few different buckets:

* Fixes that are universally good, such as removing clones and such. These are great to have!
* Fixes that the project stylistically disagrees with. For example `unwrap_or(Vec::new())` vs `unwrap_or_else(Vec::new)` or things like `for x in foo.iter()` vs `for x in &foo`. These are less great to have in my opinion.
* Fixes that I at least personally see as harmful. For example switching `match` statements to `if let` if they happen to have an empty clause, where the empty clause was written for future-proofing for future changes.  These are not great.
* Fixes that end up spraying `#[allow(clippy::...)]` annotations throughout the codebase. These to me are the worst. Each codebase tends to grow some random mish-mash of `#[allow(clippy::...)]` annotations all over the codebase, and as someone who doesn't run clippy often these are just line noise and confusing markers for future readers.

The last point is the worst in my opinion. For example skimming this PR I have absolutely no idea what `#[allow(clippy::cognitive_complexity)]` is and I feel like there's effectively zero benefit from checking it in to the code base. Cargo "solves" this with a [big block in `lib.rs`](https://github.com/rust-lang/cargo/blob/ad4122a4c8bae5558f07bc388be8f2fd870d4b80/src/cargo/lib.rs#L4-L31) to concentrate all the lints in one location, but I'm personally not a fan of that solution either. It's just moving the obscurity somewhere else.

I personally feel that clippy has failed to make itself useful to the community in this regard. The enthusiasm behind folks running clippy, matched with the default set of verbose lints, means that every project in the ecosystem is subjected to whatever style clippy happens to choose is "**the** style" as well as dealing with the pedantray of on-by-default clippy lints. AFAIK there is pretty little review of lints going into clippy and having a noisy lint on-by-default can be quite painful in the ecosystem as soon as everyone starts sending out "clippy fixes" prs for warnings that crop up.

I'm personally somewhat biased in all this in that I've never liked clippy myself. I personally wish that "clippy fixes" PRs were never sent to projects unless they ask for them. I'm certainly not the only opinion though! I wanted to lay out some of the negative experience we've had in Cargo, however, to make sure it was known.

sunfishcode commented:
I disagree with some of clippy's choices. And I can see how some of the things it does could be harmful across the ecosystem. But sometimes it finds bugs that would be difficult to find otherwise. Sometimes it improves readability. And in cases where Rust provides two equally-good ways to do something, sometimes there are advantages to having a tool like clippy make a choice so that code doesn't differ gratuitously across a codebase.

I myself don't mind some level of directives to disable clippy lints, but I can see how they cause noise. While a lib.rs list of globally disabled lints isn't ideal, it does seem like a reasonable compromise -- most people won't have to look at it most of the time.

XAMPPRocky commented:
@alexcrichton Thanks for writing about your experience! I don't have a opinion on the particular lints clippy chooses to warn about, however I disagree with the point that clippy has "failed" to be useful. For me clippy has followed a similar path to rustfmt. I disagree with a lot of formatting rustfmt chooses, however for medium–large projects I believe it's a good default.

I've seen similar complaints for both tools where authors disagree with the stylistic changes. However what I would say against that is that unless the author's have provided a style guide, you're not providing contributors with a way to follow your style, which makes contributing harder for what in my opinion is extremely low benefit.

Clippy at least for me has been very helpful at getting a grasp at understanding larger codebases like wasmtime, because it takes you through a lot of the code at different places.

I do agree that the `allow` attributes are not great though, at least not without comments explaining why they are allowed. I'll make that change with changing `unwrap_or` macro.

XAMPPRocky commented:
@sunfishcode Okay, I've updated the code with the changes to unwrap_or. There was one part, mainly around including the `level` that couldn't be reused, so I kept the old macro and had it call the new macro with that information included.

sunfishcode commented:
This looks good to me. Going forward, we should make and document a policy for clippy fixes, but for now, I think it makes sense to land this.
